research/picai/nnunet_scripts/eval.py

```diff
@@ -1,26 +1,188 @@
+    Loads images from a folder given a list of case identifiers
+
+    Args:
+        folder (str): The folder containing the images
+        case_identifiers (List[str]): A list of case identifiers for each
+            file. Typically just the filenames without the extension
+        postfixes (Optional[List[str]], optional): A list of strings to append
+            to the case identifiers when looking for files. For example
+            '_labels'. Defaults to None.
+        extensions (List[str], optional): A list of possible image extensions.
+            Defaults to [".nii.gz", ".nii", ".mha", ".mhd", ".npz", ".npy"].
+
+    Returns:
+        NDArray: A numpy array containing the images for each case identifier.
+            The first dimension will be the number of images and thos images
+            will be in the same order as was given by the case_identifiers
+            argument
+    """
+    if postfixes is None:
+        postfixes = [""]
+
+    images = []
+    for case in case_identifiers:
+        for postfix in postfixes:
+            for ext in extensions:
+                path = join(folder, case + postfix + ext)
+                if os.path.exists(path):
+                    images.append(read_image(path))
+
+    return np.stack(images)
+
+
+def get_detection_maps(probability_maps: NDArray) -> NDArray:
+    """
+    Generates detection maps from probability maps by doing lesion extraction
+
+    Args:
+        probability_maps (NDArray): A numpy array containing the predicted
+            probability maps. Should be shape (num_samples, num_classes, ...)
+
+    Returns:
+        NDArray: A numpy array of the detection maps for each class. Note that
+            a detection map is not created for the background class which is
+            assumed to be at index 0. Therefore the output shape is
+            (num_samples, num_classes-1)
+    """
+    num_classes = probability_maps.shape[1]
+
+    detection_maps = []
+    for prob_map in probability_maps:
+        cls_probs = []
+        for cls in range(1, num_classes):  # Ignore background class at idx 0
+            cls_probs.append(extract_lesion_candidates(prob_map[cls], num_lesions_to_extract=5)[0])
+        detection_maps.append(np.stack(cls_probs))
+
+    return np.stack(detection_maps)
+
+
+def get_picai_metrics(
+    detection_maps: NDArray,
+    ground_truth_annotations: NDArray,
+    case_identifiers: Optional[Iterable[str]] = None,
+    **kwargs: Any
+) -> PicaiEvalMetrics:
+    """
+    Computes the picai evaluation metrics provided the predicted lesion
+    detection maps and the ground truth annotations. Extends picai_eval to
+    allow multiclass evaluation
+
+    Args:
+        detection_maps (NDArray): The predicted lesion detection maps. Must
+            have shape (num_samples, num_lesion_classes, ...). The background
+            class should not be included in the detection maps
+        ground_truth_annotations (NDArray): The ground truth annotations. Must
```

Comment:
Yeah so the idea was the num_classes = num_lesion_classes +1 where the +1 is the background class. Hence why I started using the term num_lesion_classes instead of num_classes

Reply:
I will look to clarify this